public/app/features/logs/logsFrame.test.ts

```diff
@@ -0,0 +1,167 @@
+import { FieldType, DataFrameType, Field, Labels } from '@grafana/data';
+
+import { parseLogsFrame } from './logsFrame';
+
+function makeString(name: string, values: string[], labels?: Labels): Field {
+  return {
+    name,
+    type: FieldType.string,
+    config: {},
+    values,
+    labels,
+  };
+}
+
+function makeTime(name: string, values: number[], nanos?: number[]): Field {
+  return {
+    name,
+    type: FieldType.time,
+    config: {},
+    values,
+  };
+}
+
+function makeObject(name: string, values: Object[]): Field {
+  return {
+    name,
+    type: FieldType.other,
+    config: {},
+    values,
+  };
+}
+
+describe('parseLogsFrame should parse different logs-dataframe formats', () => {
+  it('should parse a dataplane-complaint logs frame', () => {
+    const time = makeTime('timestamp', [1687185711795, 1687185711995]);
+    const body = makeString('body', ['line1', 'line2']);
+    const severity = makeString('severity', ['info', 'debug']);
+    const id = makeString('id', ['id1', 'id2']);
+    const attributes = makeObject('attributes', [
+      { counter: '38141', label: 'val2', level: 'warning' },
+      { counter: '38143', label: 'val2', level: 'info' },
+    ]);
+
+    const result = parseLogsFrame({
+      meta: {
+        type: DataFrameType.LogLines,
+      },
+      fields: [id, body, attributes, severity, time],
+      length: 2,
+    });
+
+    expect(result).not.toBeNull();
+    if (result == null) {
+      // to make typescript happy
+      throw new Error('should never happen');
+    }
```

Comment:
You can also use the non-null assertion operator `!`.

Reply:
i hate that operator (details on demand), but you're right, in a test it should be ok. will change it.

Reply:
done.